File: arrays_and_hashing/238.py
from typing import List

# https://leetcode.com/problems/product-of-array-except-self/
# O(n), O(1)
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        ans = [1 for _ in range(len(nums))]
        # right to left
        for i in range(len(nums) - 1):
            ans[i + 1] = nums[i] * ans[i]

        # left to right
        post = 1
        for i in range(len(nums) - 1, -1, -1):
            ans[i] *= post
            post *= nums[i]
        return ans



if __name__ == "__main__":
    nums = [1, 2, 3, 4]
    print(Solution().productExceptSelf(nums))

File: arrays_and_hashing/238.cpp
#include <iostream>
#include <vector>
using namespace std;


// https://leetcode.com/problems/product-of-array-except-self/
// O(n), O(1)
class Solution {
  public:
    vector<int> productExceptSelf(vector<int>& nums) {
      vector<int> ans(nums.size(), 1);
      // left to right
      for (int i = 0; i < nums.size() - 1; i++) {
        ans[i + 1] = nums[i] * ans[i];
      }
      // right to left
      int post = 1;
      for (int i = nums.size() - 1; i >= 0; i--) {
        ans[i] *= post;
        post *= nums[i];
      }

      return ans;
    }
};


int main (int argc, char *argv[])
{
  vector<int> nums = {1,2,3,4};
  for (int n: Solution().productExceptSelf(nums)) {
    cout << n << ' ';
  }
  cout << endl;
  return 0;
}
